File: frontend/src/api/topics.ts
import client from './client';
import { Topic } from '../types/api';

export async function listTopics(): Promise<Topic[]> {
  const { data } = await client.get<Topic[]>('/api/topics');
  return data;
}

export async function createTopic(topic: Partial<Topic> & { code: string; target_tokens_per_session: number; total_token_target: number }): Promise<Topic> {
  const { data } = await client.post<Topic>('/api/topics', topic);
  return data;
}

export async function updateTopic(topicId: number, topic: Partial<Topic>): Promise<Topic> {
  const { data } = await client.put<Topic>(`/api/topics/${topicId}`, topic);
  return data;
}

File: backend/app/routers/topics.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from ..schemas.topic import TopicCreate, TopicOut, TopicUpdate
from ..security import get_current_user, get_db, require_admin
from ..services import topic as topic_service

router = APIRouter(prefix="/api/topics", tags=["topics"])


@router.get("", response_model=list[TopicOut])
def list_topics(db: Session = Depends(get_db), _: Depends = Depends(get_current_user)):
    return topic_service.list_topics(db)


@router.post("", response_model=TopicOut)
def create_topic(topic_in: TopicCreate, db: Session = Depends(get_db), admin=Depends(require_admin)):
    return topic_service.create_topic(db, topic_in)


@router.put("/{topic_id}", response_model=TopicOut)
def update_topic(topic_id: int, topic_in: TopicUpdate, db: Session = Depends(get_db), admin=Depends(require_admin)):
    return topic_service.update_topic(db, topic_id, topic_in)
